packages/instantsearch.js/src/lib/utils/render-args.ts

```diff
@@ -30,6 +30,10 @@ export function createRenderArgs(
   parent: IndexWidget
 ) {
   const results = parent.getResults()!;
+  // To make the UI optimistic, we will always render using the current state,
+  // but the previous results. This means a change will be visible immediately,
+  // regardless of the status of the network request.
+  results._state = parent.getHelper()!.state;
```

Comment:
alternative that doesn't cause the change in getServerState (but is likely slower due to the constructor being more work to instantiate multiple times per render)

```suggestion
  const parentResults = parent.getResults()!;
  const results = new algoliasearchHelper.SearchResults(
    parentResults._state,
    parentResults._rawResults
  );
  // To make the UI optimistic, we will always render using the current state,
  // but the previous results. This means a change will be visible immediately,
  // regardless of the status of the network request.
  results._state = parent.getHelper()!.state;
```

probably a couple other alternatives that don't mutate possible too, but they'll all be higher impact probably